### PR TITLE
cilium-docker: pass default cilium url when cilium-api is not provided

### PR DESCRIPTION
**Pull request by nirmoy**
Fixes: #4308

Signed-off-by: Nirmoy Das <ndas@suse.de>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4309)

### CODE REVIEW

**Comment by eloycoto** [comment]
test-me-please

File: plugins/cilium-docker/main.go
Comment (aanm) [inline plugins/cilium-docker/main.go]: `driver.NewDriver(ciliumAPI)` calls 

https://github.com/cilium/cilium/blob/1cf69804d818b88915a9c0b2b698b2c452706d2a/plugins/cilium-docker/driver/driver.go#L83

which calls
https://github.com/cilium/cilium/blob/e1850190e6e784aa1f72d554c1a3817ceaf776bd/pkg/client/client.go#L67-L76

Why is the code on `pkg/client/clieng.go` not being sufficient?

Comment (nirmoy) [reply plugins/cilium-docker/main.go]: There is no problem with pkg/client/client.go. The problem is scoppedLog is set with nil url if cilium-api isn't provided which ends up logging with no url.
e.g level=info msg="Waiting for cilium daemon to start up..." url=
https://github.com/cilium/cilium/blob/995332510c4dc08a4ab0eea81f686107c52135be/plugins/cilium-docker/driver/driver.go#L81-L83

Comment (aanm) [reply plugins/cilium-docker/main.go]: @nirmoy oh, on that case I'll defer to @raybejjani

File: plugins/cilium-docker/main.go
Comment (tgraf) [inline plugins/cilium-docker/main.go]: Code is correct. Can you move the above into a new function `DefaultSockPath() string` in `pkg/client/client.go`? So we can share the same code for this and `client.NewClient()` as well. The change itself is absolutely correct.

**Review by tgraf — CHANGES_REQUESTED** [review]

File: plugins/cilium-docker/main.go
Comment (raybejjani) [reply plugins/cilium-docker/main.go]: I think most of the handling can be implemented by using `viper.BindEnv` (e.g. https://github.com/cilium/cilium/blob/master/daemon/main.go#L306) as part of setting up the options below. That makes it consistent (however cobra/viper defines consistent) with our other parameters that are both CLI and environment variables. The default can also be set directly there, so it will be printed in the help text.

@tgraf suggested something incompatible with what I said, so he will have to overrule what I'm saying (or not) to avoid causing confusion. If tgraf really wants the new function in client.go then a compromise would be to have that function be the "default" value of the parameter, implicitly doing all the logic it does.

**Review by raybejjani — CHANGES_REQUESTED** [review]
This is a little more than a logging change. This daemon now supports passing cilium-api by env var. (unless NewDriver discards this variable after setting up the logger?)
Could you relabel the shortlog to start with `cilium-docker:` instead of `logging:`?  It might be worth adding a minor release note to this effect as well (I think that needs a `release-note/minor` tag and a quoted text block (triple \`) with `release-note` after the \`.

**Comment by nirmoy** [comment]
@raybejjani NewDefaultClient() would need such default sock path handling
https://github.com/cilium/cilium/blob/6b2b663fb1ba88b3f8d754e31d45b8016d07ca04/pkg/client/client.go#L62-L64

**Review by raybejjani — APPROVED** [review]
Well, NewDefault client could just do `NewClient(default.SockPath)` which is much more clear in what it does. Unfortunately, NewClient already handles the empty case as a special case so it's a little more complicated to fix that (there aren't that many places it is used, however). My preference would be to not do magic inside NewClient since we now have a way to get DefaultSockPath, but I don't think it's worth fixing that here...
It's also unfortunate that you had to add a similar magic handling to NewDriver, but I realise this whole PR was to get the url logged in that one place (and, also, the behaviour was already there but wasn't logged correctly).
Could you document this behaviour for both NewClient and NewDriver? If you get a chance before this PR is approved/merged.

**Comment by nirmoy** [comment]
@raybejjani @tgraf: I updated the PR let know what do you think.

[updated] added Thomas

File: plugins/cilium-docker/driver/driver.go
Comment (raybejjani) [inline plugins/cilium-docker/driver/driver.go]: The empty line detaches this comment from the function below (so it won't show up in godocs). Was this intentional? (probably not?)

**Comment by raybejjani** [comment]
@nirmoy Excellent. Thanks :) I had one small comment but I'm good.

File: plugins/cilium-docker/driver/driver.go
Comment (nirmoy) [reply plugins/cilium-docker/driver/driver.go]: oops sorry I fixed it, wasn't intentional.

Comment (raybejjani) [reply plugins/cilium-docker/driver/driver.go]: no worries :)

**Comment by raybejjani** [comment]
test-me-please

**Review by vadorovsky — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Pull request merged**
